Preserve mesh, material, and texture names in GLTF export

nikitalita opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

When serializing a Godot scene to a GLTF document, the original mesh, material, and texture names are not currently being serialized. This makes it difficult to compare exported scenes with the original and verify that GLTF exports are accurate. This changes the serialization process such that if the meshes, materials, and textures have names, we use them and we save them to the document.

fire commented:
I support this, but I am not able to review during the next week. @aaronfranke and @lyuma are around.

aaronfranke commented:
Coincidentally I just made a PR for doing this with mesh names 2 days ago: https://github.com/godotengine/godot/pull/109421

Can you make sure that the code in your PR matches this for meshes, and works the same way for material and texture names?

nikitalita commented:
> Coincidentally I just made a PR for doing this with mesh names 2 days ago: #109421
> 
> Can you make sure that the code in your PR matches this for meshes, and works the same way for surface, material, and texture names?

Ok, I rebased on top of yours and left all your changes intact, and I added my changes for materials and images.

AThousandShips commented:
4.5 has not been forked yet, the *documentation* has but not the main repo

aaronfranke committed:
Preserve mesh, material, and texture names in GLTF export

aaronfranke commented on modules/gltf/gltf_document.cpp:
```diff
@@ -4490,6 +4500,16 @@ Error GLTFDocument::_parse_texture_samplers(Ref<GLTFState> p_state) {
 	return OK;
 }
 
+inline void _set_material_texture_name(const Ref<Texture2D> &p_texture, const String &mat_name, const String &path, const String &p_suffix) {
```
Considering how `mat_name` and `p_suffix` are used together, these should be grouped. Also, parameter names must start with `p_`.

```suggestion
inline void _set_material_texture_name(const Ref<Texture2D> &p_texture, const String &p_path, const String &p_mat_name, const String &p_suffix) {
```

aaronfranke commented on modules/gltf/gltf_document.cpp:
```diff
@@ -123,12 +123,18 @@ static Ref<ImporterMesh> _mesh_to_importer_mesh(Ref<Mesh> p_mesh) {
 	for (int32_t surface_i = 0; surface_i < p_mesh->get_surface_count(); surface_i++) {
 		Array array = p_mesh->surface_get_arrays(surface_i);
 		Ref<Material> mat = p_mesh->surface_get_material(surface_i);
+		String surface_name = array_mesh.is_valid() ? array_mesh->surface_get_name(surface_i) : String();
```
```suggestion
		const String surface_name = array_mesh.is_valid() ? array_mesh->surface_get_name(surface_i) : String();
```

nikitalita commented on modules/gltf/gltf_document.cpp:
```diff
@@ -4490,6 +4500,16 @@ Error GLTFDocument::_parse_texture_samplers(Ref<GLTFState> p_state) {
 	return OK;
 }
 
+inline void _set_material_texture_name(const Ref<Texture2D> &p_texture, const String &p_path, const String &p_mat_name, const String &p_suffix) {
```
What's the difference between `inline` and `static inline`?

AThousandShips replied:
Inline doesn't guarantee namespace locality as far as I know, `static` makes sure the method isn't colliding with any other scope (like scu builds)

aaronfranke reviewed (approved):
Looks great, thank you!
